.grit/patterns/js/es6_imports.md

```diff
@@ -12,26 +12,57 @@ tags: #js, #es6, #migration, #cjs, #commonjs
 engine marzano(0.1)
 language js
 
+function transformProps($imports) {
+    $import_list = [],
+    $imports <: some bubble($import_list) {
+        or {
+            shorthand_property_identifier_pattern() as $key where $import_list += $key,
+            pair_pattern($key, $value) where $import_list += `$key as $value`
+        }
+    },
+    return join(list = $import_list, separator = ", "),
+}
+
 or {
     `const $sentry = require('@sentry/node')` => `import * as $sentry from '@sentry/node'`,
     // see https://github.com/motdotla/dotenv#how-do-i-use-dotenv-with-import
     `require("dotenv").config($config)` => `import * as dotenv from 'dotenv';\ndotenv.config($config)`,
 
-    `const { $imports } = require($source)` where {
-        $import_list = [],
-        $imports <: some bubble($import_list) {
-            or {
-                shorthand_property_identifier_pattern() as $key where $import_list += $key,
-                pair_pattern($key, $value) where $import_list += `$key as $value`
+    `const { $id: { $imports } } = require($specifier)` => `import { $id } from $specifier\nconst { $imports } = $id`,
+    `const { $imports } = require($specifier)` where { $transformed = transformProps($imports) } => `import { $transformed } from $specifier`,
+    `const $id = require($specifier).default` => `import $id from $specifier`,
+    `const $id = require($specifier).$named` => `import { $id } from $specifier` where { $id <: $named},
+    `const $id = require($specifier).$key` => `import { $key as $id } from $specifier`,
+
+    `const $id = $requireCall` where {
+        $requireCall <: contains r"require\(([^)]+)\)\.(.+)"($specifier, $rest)
+    } => `import __$id from $specifier;\nconst $id = __$id.$rest`,
+
+    `const $declarations` as $whole where {
+        $new_declarations = [],
+        $declarations <: contains bubble($new_declarations) or {
+            `$id = require($specifier).$named`,
+            `{ $id: $value } = require($specifier)`,
+            `$id = require($specifier)`
+        } where {
+            if ($named <: not undefined) {
+                if($named <: $id) {
+                  $new_declarations += `import { $id } from $specifier;`
+                } else {
+                  $new_declarations += `import { $named as $id } from $specifier;`
+                }
+            } else if ($value <: not undefined) {
+                $new_declarations += `import { $id } from $specifier;\nconst $value = $id`
+            } else {
+                $new_declarations += `import $id from $specifier;`
             }
```

Comment:
This looks like a lot of the same logic we use above for all imports - we should probably split it into a separate sub-pattern and re-use.

Reply:
Good catch, I've refactored it to use a single flow